File: app/src/main/java/com/dominikgaller/alpinebooking/booking/outbound/integration/LoggingDomainEventPublisher.java
package com.dominikgaller.alpinebooking.booking.outbound.integration;

import com.dominikgaller.alpinebooking.booking.core.domain.event.TourBookingRequested;
import com.dominikgaller.alpinebooking.booking.core.outport.DomainEventPublisher;
import org.springframework.context.ApplicationEventPublisher;

/**
 * Delegates domain event publication to Spring's {@link ApplicationEventPublisher}.
 *
 * <p>The event is published within the active transaction so that
 * {@code @TransactionalEventListener(phase = AFTER_COMMIT)} listeners receive it
 * only after the transaction has committed successfully (see ADR 0002).
 *
 * <p>SDD: See {@code documentation/ports/domain-event-publisher.outport.spec.md}
 *          and {@code documentation/adr/0002-domain-event-publication.adr.md}.
 */
public class LoggingDomainEventPublisher implements DomainEventPublisher {

    private final ApplicationEventPublisher applicationEventPublisher;

    public LoggingDomainEventPublisher(final ApplicationEventPublisher applicationEventPublisher) {
        this.applicationEventPublisher = applicationEventPublisher;
    }

    @Override
    public void publish(final TourBookingRequested event) {
        applicationEventPublisher.publishEvent(event);
    }
}

File: src/main/kotlin/com/example/contractmanagement/shared/outbound/integration/LoggingDomainEventPublisher.kt
package com.example.contractmanagement.shared.outbound.integration

import com.example.contractmanagement.shared.domain.event.DomainEvent
import com.example.contractmanagement.shared.outport.DomainEventPublisher
import org.springframework.context.ApplicationEventPublisher

/**
 * Delegates domain event publication to Spring's [ApplicationEventPublisher].
 *
 * The event is published within the active transaction so that listeners annotated
 * `@TransactionalEventListener(phase = AFTER_COMMIT)` receive it only after the
 * transaction has committed successfully (ADR 0002).
 *
 * SDD: see `documentation/ports/domain-event-publisher.outport.spec.md`.
 */
class LoggingDomainEventPublisher(
    private val applicationEventPublisher: ApplicationEventPublisher,
) : DomainEventPublisher {
    override fun publish(event: DomainEvent) {
        applicationEventPublisher.publishEvent(event)
    }
}
